### PR TITLE
feat: custom objects and records API for salesforce

### PR DESCRIPTION
**Pull request by albertyfwu**
## Test Plan

[Describe test plan here]

## Deployment instructions

[Add any special deployment instructions here]

### CODE REVIEW

File: openapi/v2/crm/components/schemas/create_update_custom_object.yaml [hunk 1]
```diff
@@ -19,6 +19,9 @@ properties:
     required:
       - singular
       - plural
+  primary_field_key_name:
```
Comment (albertyfwu) [inline openapi/v2/crm/components/schemas/create_update_custom_object.yaml:22]: both salesforce & hubspot require this. if future providers don't require it, we'll have to figure out what to do with this

File: packages/core/remotes/impl/salesforce/index.ts [hunk 2]
```diff
@@ -447,6 +458,367 @@ ${modifiedAfter ? `WHERE SystemModstamp > ${modifiedAfter.toISOString()} ORDER B
...
+            label: field.displayName,
+            type: field.fieldType === 'string' ? 'Text' : 'Number',
+            required: field.isRequired,
+          };
+
+          if (field.fieldType === 'string') {
+            return {
+              ...base,
+              // TODO: support configurations per field type
+              length: 255,
```
Comment (albertyfwu) [inline packages/core/remotes/impl/salesforce/index.ts:538]: if we want to not use defaults here, we'd need to expand our API

File: packages/core/remotes/impl/salesforce/index.ts [hunk 3]
```diff
@@ -447,6 +458,367 @@ ${modifiedAfter ? `WHERE SystemModstamp > ${modifiedAfter.toISOString()} ORDER B
...
+            };
+          } else {
+            return {
+              ...base,
+              // TODO: support configurations per field type
+              precision: field.fieldType === 'number' ? 18 : undefined,
```
Comment (albertyfwu) [inline packages/core/remotes/impl/salesforce/index.ts:544]: if we want to not use defaults here, we'd need to expand our API

File: packages/core/remotes/impl/salesforce/index.ts [hunk 4]
```diff
@@ -447,6 +458,348 @@ ${modifiedAfter ? `WHERE SystemModstamp > ${modifiedAfter.toISOString()} ORDER B
...
+    const metadata = await this.#client.metadata.read('CustomObject', id);
+    return toCustomObject(metadata);
+  }
+
+  public override async createCustomObject(params: CustomObjectCreateParams): Promise<string> {
+    if (!params.name.endsWith('__c')) {
+      // Salesforce doesn't actually enforce this, and will just append __c.
+      // However, we want to enforce this to avoid confusion when using
+      // the custom object name in other places.
+      throw new Error('Custom object name must end with __c');
```
Comment (asdfryan) [inline packages/core/remotes/impl/salesforce/index.ts:478]: This should be a 4xx and not a 5xx right?

File: packages/core/remotes/impl/salesforce/index.ts [hunk 5]
```diff
@@ -447,6 +458,348 @@ ${modifiedAfter ? `WHERE SystemModstamp > ${modifiedAfter.toISOString()} ORDER B
...
+            required: field.isRequired,
+          };
+
+          if (field.fieldType === 'string') {
+            return {
+              ...base,
+              // TODO: support configurations per field type
+              length: 255,
+            };
+          } else {
```
Comment (asdfryan) [inline packages/core/remotes/impl/salesforce/index.ts:540]: Nit: else is redundant

File: openapi/v2/crm/components/schemas/create_update_custom_object.yaml (hunk 1 above)
Comment (asdfryan) [reply openapi/v2/crm/components/schemas/create_update_custom_object.yaml:22]: Can you add a description?

File: packages/core/remotes/impl/salesforce/index.ts (hunk 3 above)
Comment (asdfryan) [reply packages/core/remotes/impl/salesforce/index.ts:544]: We should document this in the API spec

File: packages/core/remotes/impl/salesforce/index.ts (hunk 2 above)
Comment (asdfryan) [reply packages/core/remotes/impl/salesforce/index.ts:538]: We should document this in the API spec

File: packages/core/remotes/impl/salesforce/index.ts [hunk 6]
```diff
@@ -447,6 +458,372 @@ ${modifiedAfter ? `WHERE SystemModstamp > ${modifiedAfter.toISOString()} ORDER B
...
+      throw new Error('Custom object id must end with __c');
+    }
+
+    const metadata = await this.#client.metadata.read('CustomObject', id);
+    return toCustomObject(metadata);
+  }
+
+  public override async createCustomObject(params: CustomObjectCreateParams): Promise<string> {
+    if (!params.name.endsWith('__c')) {
+      // Salesforce doesn't actually enforce this, and will just append __c.
```
Comment (tomkit) [inline packages/core/remotes/impl/salesforce/index.ts:475]: Are these documented in openapi?

Comment (albertyfwu) [reply packages/core/remotes/impl/salesforce/index.ts:475]: no i'll add it

**Pull request merged**